### PR TITLE
Support expansion of all path configuration options

### LINKED ISSUE #7272: Feature: Do expansion on all path options in configuration
Previously issues #6508 and #7259 added home directory and environment variable expansion to the `cache_dir` option in the configuration file. It might make sense to the the same for all the other options also referring to paths, these would include: `files`, `mypy_path`, `custom_typeshed_dir` and `python_executable`.

Some review may be necessary to determine if this is not appropriate for any of these options.

### PR DESCRIPTION
**Pull request by hetmankp**
This closes #7272; all options in the configuration file that deal with paths will be expanded for the user home directory and environment variables.

### CODE REVIEW

File: mypy/config_parser.py [hunk 1]
```diff
@@ -63,8 +64,9 @@ def split_and_match_files(paths: str) -> List[str]:
     'python_version': parse_version,
     'strict_optional_whitelist': lambda s: s.split(),
     'custom_typing_module': str,
-    'custom_typeshed_dir': str,
-    'mypy_path': lambda s: [p.strip() for p in re.split('[,:]', s)],
+    'custom_typeshed_dir': lambda s: os.path.expandvars(os.path.expanduser(s)),
```
Comment (msullivan) [inline mypy/config_parser.py]: How about defining a function that does the two expansions together?

**Comment by hetmankp** [comment]
Ok, let me know if that's what you had in mind.

**Comment by msullivan** [comment]
Yeah!

File: mypy/config_parser.py [hunk 2]
```diff
@@ -63,8 +71,8 @@ def split_and_match_files(paths: str) -> List[str]:
     'python_version': parse_version,
     'strict_optional_whitelist': lambda s: s.split(),
     'custom_typing_module': str,
-    'custom_typeshed_dir': str,
-    'mypy_path': lambda s: [p.strip() for p in re.split('[,:]', s)],
+    'custom_typeshed_dir': expand_path,
+    'mypy_path': lambda s: [expand_path(p.strip()) for p in re.split('[,:]', s)],
     'files': split_and_match_files,
     'quickstart_file': str,
     'junit_xml': str,
```
Comment (Mattwmaster58) [inline mypy/config_parser.py]: Would enabling environment variable interpolation for this config key be as simple as this?
```suggestion
    'junit_xml': expand_path
```

Comment (hetmankp) [reply mypy/config_parser.py]: > Would enabling environment variable interpolation for this config key be as simple as this?

Yes. Probably should have done this for `junit_xml` and `quickstart_file` as well.

Comment (Mattwmaster58) [reply mypy/config_parser.py]: Alright, I'll include that in my PR #8479

**Pull request merged**